src/Stratis.Bitcoin/P2P/Protocol/Behaviors/EnforcePeerVersionCheckBehavior.cs

```diff
@@ -0,0 +1,111 @@
+
+        /// <summary>
+        /// Set to <c>true</c> if the attached peer callbacks have been registered and they should be unregistered,
+        /// <c>false</c> if the callbacks are not registered.
+        /// </summary>
+        private bool callbacksRegistered;
+
+        /// <summary>
+        /// Initializes an instance of the object for outbound network peers.
+        /// </summary>
+        /// <param name="chainIndexer">The chain of blocks.</param>
+        /// <param name="nodeSettings">User defined node settings.</param>
+        /// <param name="network">Specification of the network the node runs on - regtest/testnet/mainnet.</param>
+        /// <param name="loggerFactory">Factory for creating loggers.</param>
+        public EnforcePeerVersionCheckBehavior(ChainIndexer chainIndexer,
+            NodeSettings nodeSettings,
+            Network network,
+            ILoggerFactory loggerFactory)
+        {
+            this.chainIndexer = chainIndexer;
+            this.nodeSettings = nodeSettings;
+            this.network = network;
+            this.loggerFactory = loggerFactory;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName, $"[{this.GetHashCode():x}] ");
+        }
+
+        private Task OnMessageReceivedAsync(INetworkPeer peer, IncomingMessage message)
+        {
+            int enforceMinProtocolVersionAtBlockHeight = this.network.Consensus.Options.EnforceMinProtocolVersionAtBlockHeight;
+            bool enforcementRequired = enforceMinProtocolVersionAtBlockHeight > 0;
+            if (!enforcementRequired)
+                return Task.CompletedTask;
+
+            bool enforcedAlready = this.nodeSettings.MinProtocolVersion >= this.network.Consensus.Options.EnforcedMinProtocolVersion;
+            if (enforcedAlready)
+                return Task.CompletedTask;
+
+            bool enforcementHeightReached = this.chainIndexer.Height >= enforceMinProtocolVersionAtBlockHeight;
+            if (enforcementHeightReached)
+            {
+                this.logger.LogDebug("Changing the minumum supported protocol version from {0} to {1}.", this.nodeSettings.MinProtocolVersion, this.network.Consensus.Options.EnforcedMinProtocolVersion);
+                this.nodeSettings.MinProtocolVersion = this.network.Consensus.Options.EnforcedMinProtocolVersion;
+            }
+
+            // The statement below will close connections in case the this.NodeSettings.MinProtocolVersion has changed during node execution.
+            if (peer.PeerVersion.Version < this.nodeSettings.MinProtocolVersion)
+            {
+                this.logger.LogError("Unsupported client version, dropping connection.");
+                this.AttachedPeer.Disconnect("Peer is using unsupported client version");
+            }
+
+            return Task.CompletedTask;
+        }
+
+        [NoTrace]
+        protected override void AttachCore()
+        {
+            if (this.AttachedPeer != null)
```

Comment:
Here you ensure that attached peer is null and below you are trying to access it
that will lead to null ref exception

Reply:
The check was ensuring that the AttachedPeer is NOT null but I have removed it as it is checked in the calling method.